### PR TITLE
tanjiro: wall-shear magnitude+direction decomposition for tau_y/z

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The wall-shear vector `(tau_x, tau_y, tau_z)` is predicted and supervised in Cartesian coordinates, where tau_y and tau_z are 3-4× smaller than tau_x (longitudinal flow dominance in DrivAerML cars). Predicting small numbers in absolute coordinates is inherently hard — a 5% relative error on tau_y is a much smaller absolute deviation than a 5% relative error on tau_x, but both contribute equally to MSE loss.

**Hypothesis:** Decomposing the wall-shear vector into **magnitude + unit-direction** (or equivalently into spherical coordinates: magnitude r, elevation θ, azimuth φ) and predicting direction separately from magnitude will reduce the tau_y/z relative L2 error. The physical insight: the model will get an explicit representation of "which way is the shear pointing" instead of having to learn tau_y/z as the residual of a large tau_x minus a slightly-off prediction.

The specific change: **output 4 surface channels** — `(tau_magnitude, tau_x_normalized, tau_y_normalized, tau_z_normalized)` where the normalized direction is a unit vector `tau / |tau|` and tau_magnitude = `||tau||`. Reconstruct Cartesian tau for loss calculation: `tau_x = tau_magnitude * tau_x_normalized`, etc.

**Why this should help:**
1. tau_y/z error is dominated by direction error (the model predicts roughly the right magnitude but wrong cross-flow/spanwise direction). Separating magnitude supervision from direction supervision allows the model to learn both from dedicated feature channels.
2. The unit-direction channels have values near `[1.0, ~0.1, ~0.1]` for most DrivAerML geometries (longitudinal flow). This is a much more balanced regression target.
3. No additional parameters — just output head restructuring + loss recomposition.
4. This is orthogonal to all in-flight interventions (Huber loss, GradNorm, curriculum, tangent-frame inputs, DropPath).

**Related prior work:** fern PR #11 (tangential wall-shear projection loss, which projects the shear onto surface tangent vectors) targeted a related idea but from the loss side. This PR attacks the **output parameterization** — how the model represents and predicts the shear vector itself.

### Step 0 — Verify there's no `--use-magnitude-direction-wallshear` flag already

```bash
python train.py --help | grep -i magnitude
```

If it exists, report back before implementing.

### Step 1 — Add the flag

```python
parser.add_argument(
    "--wallshear-magnitude-direction",
    action="store_true",
    default=False,
    help="Decompose wall-shear target into magnitude + unit-direction for output head. "
         "Model predicts (tau_mag, tau_x_hat, tau_y_hat, tau_z_hat); Cartesian tau reconstructed for loss.",
)
```

### Step 2 — Modify the surface output head behavior

When `--wallshear-magnitude-direction` is set, the 4-channel surface output is reinterpreted as:
- Channel 0: `surface_pressure` (unchanged)
- Channel 1: `tau_magnitude = ||tau||` (non-negative; clamp to 0 at eval)
- Channels 2, 3, 4: `tau_unit = tau / ||tau||` (unit direction vector)

The model still uses the same 4-output architecture (single linear projection 512→4 for surface_p + tau_x + tau_y + tau_z). **Just rename these outputs conceptually in the loss code — no architecture change needed.**

Specifically:
```python
# At loss computation time, when wallshear_magnitude_direction=True:
# pred[:, 1] = pred_tau_mag   (predicted magnitude)
# pred[:, 2:5] = pred_tau_dir  (predicted direction, un-normalized)
# target[:, 1] = target_tau_mag = torch.norm(target_tau, dim=-1)
# target[:, 2:5] = target_tau_dir = target_tau / (target_tau_mag.unsqueeze(-1) + 1e-8)
# Loss: MSE on tau_mag + MSE on tau_dir (unit vector, no normalization needed at loss time)
# Metric reconstruction: pred_tau_cartesian = pred_tau_mag * normalize(pred_tau_dir)
```

Important: **at eval/metric time** reconstruct Cartesian tau for metric computation:
```python
# pred_tau_mag = pred[:, 1].clamp(min=0)
# pred_tau_dir = F.normalize(pred[:, 2:5], dim=-1)
# pred_tau_cartesian = pred_tau_mag.unsqueeze(-1) * pred_tau_dir
```

The `surface_pressure_rel_l2_pct`, `wall_shear_rel_l2_pct`, `wall_shear_x/y/z_rel_l2_pct`, `abupt_axis_mean_rel_l2_pct` metrics must all be computed on the reconstructed Cartesian outputs so results are directly comparable.

### Step 3 — Loss weight considerations

The magnitude channel and direction channels have different scales. Start with unit weight on both. If the model doesn't converge cleanly, try `--wallshear-direction-weight 0.5` to down-weight the direction loss relative to magnitude.

### Step 4 — Run the control vs treatment comparison

**Run both arms in parallel** on 2 GPUs. Each arm: same config as SOTA, single-GPU (4 GPUs available — use 2 for the comparison, leave 2 idle as spare):

**Arm A — Control (Cartesian, no decomposition):**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --lr 5e-4 --weight-decay 1e-4 \
  --batch-size 4 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 --lr-warmup-steps 500 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --volume-loss-weight 2.0 \
  --validation-every 1 --epochs 3 \
  --wandb-group tanjiro-wallshear-magdir-r0 \
  --wandb-name tanjiro/magdir-A-control \
  --kill-thresholds "3000:val_primary/abupt_axis_mean_rel_l2_pct<25" \
  --agent tanjiro &
```

**Arm B — Magnitude+Direction decomposition:**
```bash
CUDA_VISIBLE_DEVICES=1 python train.py \
  --lr 5e-4 --weight-decay 1e-4 \
  --batch-size 4 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 --lr-warmup-steps 500 \
  --wallshear-y-weight 2.0 --wallshear-z-weight 2.0 \
  --volume-loss-weight 2.0 \
  --wallshear-magnitude-direction \
  --validation-every 1 --epochs 3 \
  --wandb-group tanjiro-wallshear-magdir-r0 \
  --wandb-name tanjiro/magdir-B-magdir \
  --kill-thresholds "3000:val_primary/abupt_axis_mean_rel_l2_pct<25" \
  --agent tanjiro &
```

### Step 5 — Reporting

Report ep1/ep2/ep3 val metrics for both arms. Key metrics to report per arm:
- `val_primary/abupt_axis_mean_rel_l2_pct` (primary)
- `val_primary/wall_shear_y_rel_l2_pct` and `val_primary/wall_shear_z_rel_l2_pct` (hypothesis test — these should improve most)
- `val_primary/wall_shear_x_rel_l2_pct` (sanity — should be unaffected or slightly better)
- `val_primary/surface_pressure_rel_l2_pct` (sanity — should be unaffected)

**Decision criteria:**
- Arm B val_abupt < Arm A at ep2+ AND wall_shear_y/z improves by ≥0.5pp → promising, launch full 8-GPU Lion run for merge
- Arm B diverges or NaN → close
- Arm B val_abupt ≥ Arm A with no per-channel improvement → close

## Baseline

Current best: PR #311 (edward, STRING-separable learnable PE), W&B run `gcwx9yaa`

| Metric | val | test | AB-UPT target |
|---|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **7.546** | **8.771** | — |
| `surface_pressure_rel_l2_pct` | — | 4.485 | 3.82 |
| `wall_shear_rel_l2_pct` | — | 8.227 | 7.29 |
| `volume_pressure_rel_l2_pct` | — | 12.438 | 6.08 |
| `wall_shear_x_rel_l2_pct` | — | 7.253 | 5.35 |
| `wall_shear_y_rel_l2_pct` | — | 9.233 | 3.65 |
| `wall_shear_z_rel_l2_pct` | — | 10.449 | 3.63 |

**Merge bar: val_abupt < 7.546%** (only achievable via full 8-GPU Lion run after a promising single-GPU screen).

The tau_y/z gaps (2.53× and 2.88× AB-UPT) are the primary research target. If this direction shows ≥0.5pp improvement in the single-GPU screen, it will be worth a full-budget 8-GPU Lion run.

Note: current `yi` DDP infra requires PR #355 (emma, DDP infrastructure fix). If DDP is not yet available on this pod's `train.py`, the single-GPU screen in Step 4 is the right starting point.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #369 because ADVISOR: Closing immediately — this hypothesis has already been tested twice and both failed decisively.

- **PR #132 (violet)**: Decoupled wall-shear magnitude + direction with cosine direction loss — abupt +12.7% to +17.1% worse than baseline. tau_y/z got 23-35% WORSE. Root cause: cosine direction loss gradient → 0 as alignment improves (sin(θ) factor), so y/z error provides negligible gradient once the model learns the dominant tau_x direction. Also: unit vector reformulation doesn't preferentially upweight small-magnitude y/z axes — they remain numerical noise inside the unit vector.

- **PR #200 (emma)**: Log-magnitude + cosine direction loss — all 4 arms (α=0.3/0.5/0.7/1.0) failed catastrophically. ep1 abupt 24–84% vs baseline 10.69%. Root cause: ∂log(|τ|+ε)/∂τ → 1/(|τ|+ε) blows up near zero (gradients exceeded 8.2×10¹⁸).

The proposed MSE-on-magnitude + MSE-on-unit-direction variant avoids the log-gradient explosion but not the fundamental issue: for DrivAerML cars, the unit direction vector is approximately [0.98, 0.1, 0.1] (streamwise-dominated). MSE on the unit vector components is still dominated by the tau_x_normalized ≈ 0.98 direction; tau_y_normalized and tau_z_normalized are still small-number targets that provide weak signal relative to the dominant axis.

Both prior PRs confirm this approach class is a dead end for the tau_y/z gap. Assigning tanjiro a fresh hypothesis instead..
